doc/developer_guide/developers.rst

```diff
@@ -530,6 +530,67 @@ DataAccess class i.e. the `_field_write_arguments()` and
 `_field_read_arguments()` methods, both of which are found in the
 `Arguments` class.
 
+Variable Accesses
+=================
+
+Especially in the NEMO API, it is not possible to rely on pre-defined
+kernel information to determine dependencies between loops. So an additional,
+somewhat lower-level API is being implemented that can be used to determine
+variable accesses (READ and WRITE), which is based on the PSyIR information only,
+not on any kernel information. The information about all variable usage
+of a node can be gathered by creating an object of type
+``psyclone.core.access_info.VariablesAccessInfo``, and then calling
+the function ``reference_accesses`` for the node:
+
+.. autofunction:: psyclone.psyGen.Node.reference_accesses
+
+.. autoclass:: psyclone.core.access_info.VariablesAccessInfo
+    :members:
+
+This class collects information for each variable used in the tree
+starting with the given node. For each variable an instance of
+``psyclone.core.access_info.VariableAccessInfo`` is created, which collects
+all accesses for that variable using ``psyclone.config.access_info.AccessInfo``
+instances:
+
+.. autoclass:: psyclone.core.access_info.VariableAccessInfo
+    :members:
+
+.. autoclass:: psyclone.core.access_info.AccessInfo
+    :members:
+
+Here a simple example on how to use this API. This is from the
+``psyclone.psyGen.OMPParallelDirective`` (so ``self`` is an instance of this
+transformation), and this code is used to determine a list with all scalar
+variables that must be declared as thread-private::
+
+  var_accesses = VariablesAccessInfo()
+  self.reference_accesses(var_accesses)
+  for var_name in var_accesses.get_all_vars():
+      accesses = var_accesses.get_varinfo(var_name).get_all_accesses()
+      # Ignore variables that have indices, we only look at scalar
+      # This tests only the first access, the assumption is that all
+      # accesses are either using an array access, or none.
+      if accesses[0].get_indices() is not None:
```

Comment:
While I totally agree,I am not sure if the SymbolTable is ready to be used ... and how are we going to handle the case if we should not have any declarations? Not sure if the latter case could actually happen in reality (it will affect all test cases, since I never declared anything there ;) ). I'll bring this up in our telco tonight.

Reply:
I had a go at this issue, and I am not sure if I might have misunderstood :)  I tried to add a symbol table to the VariablesAccessesInfo class, and then add a function 'is_array()' - but it somehow feels wrong. Imho this class should really only store the information, not the interpretation - the script using the access information should decide what to use (e.g. symbol table or if there is an index in the access). 

So I re-read the ticket, I am wondering if you actually only meant: in this example in the docs we should use the symbol table (if available)? Here how I have rewritten the example code in the docs:
```
  for var_name in var_accesses.all_vars:
      accesses = var_accesses[var_name].all_accesses
      # Ignore variables that are arrays, we only look at scalar ones.
      # If we do have a symbol table, use the shape of the variable
      # to determine if the variable is scalar or not
      if symbol_table:
          if len(symbol_table.lookup(var_name).shape) > 0:
              continue
      else
          # There is no symbol table, check instead if the first access of
          # the variable has indices, and assume it is an array if it has:
          if accesses[0].indices is not None:
              continue

```

Is just the change of the example in the documentation what you wanted me to change? Or do you really want to have support for an optional symbol table inside the variable access information?

Reply:
I think I meant in the specific example and I'm happy with your suggested change. I don't think we should be adding a SymbolTable to the `VariablesAccessInfo` class. I think probably what is needed is that each Symbol in the PSyIR contain a ptr to its entry in the SymbolTable (if any) but that's not something for this Issue.

Reply:
Puh - glad to hear that (and glad that I stopped adding the SymbolTable :grin:). I've added this to the example.